troubleshoot_data.py

```diff
@@ -0,0 +1,434 @@
+        log.write(ctx, "Combined JSON matches: {}".format(result['combiJson_check']), True)
+
+    log.write(ctx, "", True)
+
+
+def collect_troubleshoot_data_packages(ctx, requested_package, write_stdout):
+    data_packages = []
+
+    if requested_package == 'None':
+        # Retrieve all data packages
+        all_packages = find_data_packages(ctx, write_stdout)
+        if not all_packages:
+            log.write(ctx, "collect_troubleshoot_data_packages: No packages found.", write_stdout)
+            return None
+
+        data_packages = all_packages
+    else:
+        # Get full path of the given package
+        full_package_path = find_full_package_path(ctx, requested_package, write_stdout)
+
+        if not full_package_path:
+            log.write(ctx, "collect_troubleshoot_data_packages: Data package '{}' cannot be found.".format(requested_package), write_stdout)
+            return None
+
+        data_packages.append(full_package_path)
+
+    return data_packages
+
+
+def batch_troubleshoot_published_data_packages(ctx, requested_package, log_file, offline, api_call):
+    """
+    Troubleshoots published data packages.
+
+    :param ctx:               Context that combines a callback and rei struct.
+    :param requested_package: A string representing a specific data package path or all packages with failed publications.
+    :param log_file:          A boolean representing to write results in log.
+    :param offline:           A boolean representing whether to perform all checks without connecting to external servers.
+    :param api_call:          Boolean of whether this is run by a script or api test.
+
+    :returns: A dictionary of dictionaries providing the results of the job.
+    """
+    write_stdout = not api_call
+    # Check permissions - rodsadmin only
+    if user.user_type(ctx) != 'rodsadmin':
+        log.write(ctx, "User is not rodsadmin", write_stdout)
+        return {}
+
+    data_packages = collect_troubleshoot_data_packages(ctx, requested_package, write_stdout)
+    if not data_packages:
+        return {}
+    schema_cache = {}
+    results = {}
+
+    # Troubleshooting
+    for data_package in data_packages:
+        log.write(ctx, "Troubleshooting data package: {}".format(data_package), write_stdout)
+        result = {}
+        # Cannot check the metadata as technicaladmin
+        if not api_call:
+            schema_check_dict = vault_metadata_matches_schema(ctx, data_package, schema_cache, "troubleshoot-publications", write_stdout)
+            result['schema_check'] = schema_check_dict['match_schema'] if schema_check_dict else False
+
+        result['no_missing_AVUs_check'], result['no_unexpected_AVUs_check'] = check_print_data_package_system_avus(ctx, data_package, write_stdout)
+
+        # Do not check datacite in offline mode
+        if not offline:
+            result['versionDOI_check'], result['baseDOI_check'] = check_datacite_doi_registration(ctx, data_package, write_stdout)
```

Comment:
Should we have a flag for the datacite doi check separate from the offline flag?